### PR TITLE
20140: [Feature] DistributedData DataDeleted msg should also provide requestor context

### PR DESCRIPTION
**Pull request by kelebra**
Could you please advice on a proper place where to add test cases for this? I just do not feel like it will be right to create a separate test class for that. Thanks.

### CODE REVIEW

**Comment by kelebra** [comment]
Refs #20527

**Comment by ktoso** [comment]
OK TO TEST

**Comment by kelebra** [comment]
Failure caused by:
```bash
java.lang.RuntimeException: akka-distributed-data-experimental: Binary compatibility check failed!
```
From the issue discussion:
> we are allowed to break binary compatibility here, because it's an experimental module

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 1]
```diff
@@ -408,9 +408,9 @@ object Replicator {
   sealed trait DeleteResponse[A <: ReplicatedData] {
     def key: Key[A]
   }
-  final case class DeleteSuccess[A <: ReplicatedData](key: Key[A]) extends DeleteResponse[A]
-  final case class ReplicationDeleteFailure[A <: ReplicatedData](key: Key[A]) extends DeleteResponse[A]
-  final case class DataDeleted[A <: ReplicatedData](key: Key[A])
+  final case class DeleteSuccess[A <: ReplicatedData](key: Key[A], request: Option[Any] = None) extends DeleteResponse[A]
+  final case class ReplicationDeleteFailure[A <: ReplicatedData](key: Key[A], request: Option[Any] = None) extends DeleteResponse[A]
+  final case class DataDeleted[A <: ReplicatedData](key: Key[A], request: Option[Any] = None)
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: shouldn't `request: Option[Any]` parameter be added to the `Delete` message?

Comment (kelebra) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: From the issue itself (what was requested):
```
DataDeleted(key: Key, request: Some[Any])
DeleteSuccess(key: Key, request: Some[Any])
ReplicationDeleteFailure(key: Key, request: Some[Any])
```
Maybe you meant ```DeleteResponse``` to have it?

Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: You send a `Delete` message and get a `DeleteSuccess` back as reply. The feature is that you can pass a request object in the `Delete` message same object is passed in the reply. Perhaps you should add a tests to verify that it works?

**Comment by patriknw** [comment]
@huntc It was an oversight in the API, the request context was missing from the `Delete` message and friends. Want to check that it's alright for your usage that we change this in a non-binary compatible way?

**Comment by huntc** [comment]
I'm pretty sure we don't use Delete - once deleted a key can't be used again right? Ido t think we do that...

**Comment by patriknw** [comment]
yes, that is how delete works

**Comment by huntc** [comment]
@patriknw Confirming that we do not use `Delete` within ConductR. Breaking binary compatibility in this regard won't hurt us. Thanks for asking.

**Comment by patriknw** [comment]
@huntc Thanks for confirming.

Is the following correct, for my understanding of how you use distributed data? You don't use it in any of the end-user libraries, but only in ConducR itself. That means that it will never be used with another Akka version than what you compiled it with, and therefore breaking binary compatibility would not be an issue (apart from that you might have to source code adjustments if we break source compatibility).

**Comment by patriknw** [comment]
Closing due to inactivity. New attempt is welcome.

**Pull request closed without merging**